### ISSUE TITLE
Provide wrappers for tool binaries instead of symlink

### ISSUE THREAD

**Issue by kateinoigakukun** [swiftlang/swiftly#52]
The design document says:

> The ~/.local/bin directory would include symlinks pointing to the bin directory of the "active" toolchain, if any.
> This is all very similar to how rustup does things, but I figure there's no need to reinvent the wheel here.

https://github.com/swift-server/swiftly/blob/4d718c0621bdfd5835507565896ef9d28b390659/DESIGN.md#installation-of-a-swift-toolchain

But it looks like `rustup` does more things, known as [`Proxies`](https://rust-lang.github.io/rustup/concepts/proxies.html), to select the active toolchain based on `rust-toolchain.toml` or `cargo +nightly` style at the actual command invocation time.

Of course, I think it makes a lot of sense to start by limiting the toolchain selection only globally by creating symlinks at installation time (as swiftly does now).
But I'd like to raise up the future direction discussion of shims just for the record. (Sorry if it's already discussed somewhere)

**Comment by adam-fowler** [comment]
I would rather we did not use proxy shell scripts. The swiftly swift install is non-standard in that the binaries are referenced from a folder outside the actual swift toolchain folder. With symbolic links it is very easy to find the true bin folder (and thus the toolchain folder) for the current installation by following the symbolic links. If we change this to wrapper shell scripts this process becomes a lot harder and more brittle as we have to parse these shell scripts and rely on them following a certain structure.

Being able to reach the actual true swift installation is needed by many tools that makes use of the swift toolchain eg the VSCode swift extension.

**Comment by kateinoigakukun** [comment]
Okay, I understand the problem. Thank you for elaborating clearly!

However, I think avoiding proxy loses a lot of opportunities to provide practical features (e.g. select version based on `.swift-version` file, select a specific version locally (not globally) )

IMHO, for the `liblldb` issue, the proper way would be to ask `swift` command to provide the canonical toolchain directory by something like `swift --print-toolchain-path`. Note that the driver doesn't have the option now but we can have some workarounds for now by getting the current executable (lldb) by REPL mode:
```
 $ echo "print(CommandLine.arguments[0])" | swift repl 2> /dev/null` 
/home/katei/.local/share/swiftly/toolchains/main-snapshot-2022-06-08/usr/bin/repl_swift
```

**Comment by kateinoigakukun** [comment]
@patrickfreed  Do you have any thoughts on this?

**Comment by patrickfreed** [comment]
I think the idea of using proxies rather than simple symlinks is very interesting (and TIL rustup uses them, that was an oversight by me in the design), but I'm hesitant to introduce them until we have a specific feature in mind that we want. 

I think you've pointed out the two main use-cases, namely a local specification of the in-use toolchain (e.g. a `.swift-version` file or some such) or some alternative one-off syntax for selecting a version (similar to `+nightly` in cargo). 

Regarding the former, it was an explicit decision to omit that feature for now for and to see if it turned out to be something that's important to users. 

Regarding the latter, swiftly won't be as ubiquitous in the Swift ecosystem as rustup is in the Rust ecosystem, so I think introducing new syntax to the `swift` executable might not be the best idea (you can imagine users writing scripts that work with swiftly-installed toolchains but not manually downloaded ones). We might also run into issues supporting such syntax on macOS. Instead, I think it might be better to introduce a new command in swiftly itself for that kind of functionality (see #22), which would mean that shims wouldn't be required.

So in summary, I think its an interesting idea that we might want to consider pursuing in the future, but I'd rather let it be done as part of implementing some other user-visible feature rather than in anticipation of such a feature. I will admit that switching from symlinks to shims in a later release of swiftly will require a bit more work from us (it should be invisible to the user at least), but I think I'm okay with that.

Thanks for bringing this up!

**Comment by kateinoigakukun** [comment]
Okay, it makes sense to me.

**Comment by patrickfreed** [comment]
If https://github.com/apple/swift/issues/70932 (#92) doesn't get resolved, we may need to consider doing this to ensure swiftly continues to work with the new driver. Ideally, we won't need to do anything here though--I agree with @adam-fowler's assessment that this really is a bug.

**Comment by kkebo** [comment]
FYI

https://github.com/apple/swift-driver/pull/1583

**Comment by cmcgee1024** [comment]
@kateinoigakukun we've changed swiftly to use proxies recently to support workflows, like selecting a toolchain automatically based on what's in the `.swift-version` file. Is this the direction that you were thinking when you originally raised this? If so, maybe we can close this now?

**Comment by kateinoigakukun** [comment]
Thanks @cmcgee1024 !
